### ISSUE TITLE
Creating several partitions on a single file

### ISSUE THREAD

**Issue by chdoig** [dask/dask#287]
**tl;dr:** Is it/Would it be possible to handle parallelism for processing a single large file with dask?

Use case example: Processing wikipedia articles dump.

I'm trying to reproduce a tokenization task from gensim with dask:  http://radimrehurek.com/gensim/wiki.html#preparing-the-corpus

*Convert the articles to plain text (process Wiki markup) and store the result as sparse TF-IDF vectors. In Python, this is easy to do on-the-fly and we don’t even need to uncompress the whole archive to disk. There is a script included in gensim that does just that, run:*

    $ python -m gensim.scripts.make_wiki

That script uses the class WikiCorpus, which uses multiprocessing the following way:

https://github.com/piskvorky/gensim/blob/f1a904d61cdea8e44681342678a5eefd178d7b12/gensim/corpora/wikicorpus.py#L291-L308

The file is currently ~11GB compressed.

This problem could also help this library to extract wiki dumps:
https://github.com/attardi/wikiextractor

where they were also asking for help on a multi-threading/multi-processing that same task: https://github.com/attardi/wikiextractor/issues/4

I'd like to tackle this problem with dask but, I don't know how/where I should start. Is it possible to create several bags after parsing a single file? Similarly to what gensim in that script does?

cc: @mrocklin

**Comment by mrocklin** [comment]
We can parallelize a single file if that file is essentially line delimited.  We do this with `dd.read_csv` using this function  https://github.com/ContinuumIO/dask/blob/master/dask/utils.py#L134-L176

I'll try to take a look at this in the next couple of days.

**Comment by chdoig** [comment]
Thanks!

In this case the file is not line delimited. It uses xml tags <page></page> to delimit pages. Both solutions, essentially use regex to parse the file:

https://github.com/piskvorky/gensim/blob/f1a904d61cdea8e44681342678a5eefd178d7b12/gensim/corpora/wikicorpus.py#L173-L199

https://github.com/attardi/wikiextractor/blob/c15d93c40aee54e2875aa7fd598fa198cd492899/extractPage.py#L47-L77

I was thinking maybe of a mechanism that could create several partitions after parsing, e.g:

```
items = db.from_filename('foo.xml.bz2').parse_file(parsing_function)
```
and them process those items like another dask.bag with several partitions.

**Comment by mrocklin** [comment]
We would need some way to cleanly partition the data in the file without
doing much work ahead of time.  I don't know if xml supports this kind of
clean partitioning.  Actually, I'm not sure if xml even supports streaming
cleanly.

**Comment by mrocklin** [comment]
Partitioning after the fact is possible but unpleasant because it would kill streaming and would require us to send data from one process back to the central process (undergoing a serialization-deserialization cycle) and then send those parts off to other workers (again suffering serialization-deserialization).

It's entirely possible that this would be significantly slower than doing it in a single process.  My inclination here is not to try to break up unpartitionable formats like full XML and to stick to more partitionable formats like line-delimited json, csv, etc..

Thoughts?

I'll implement a fragment operation so you can try it out.  I wouldn't expect much though.

**Comment by chdoig** [comment]
I had a good conversation about this with @eriknw.

Use case: I have a large file, that I know how to parse to get items that are not line delimited. I have a function that can yield those items. I want to be able to easily express: have a worker that takes x of those items and does some computation with them, another worker comes and takes the next x items and does some computation with them...

Maybe something like:

```python
# No dask, I create the extract_pages function
texts = ((text, title, pageid) for title, text, pageid in extract_pages(bz2.BZ2File(fname)))

# dask
import dask.bag as db
db.from_generator(texts).map(...).compute()
```

or something like

```python
# No dask, provided by user, custom to file format
def read_file(f):
    while True:
        record = f.read_record()
        t = record.payload.read()
        yield t

# dask
b = (db.from_sequence([fname])
     .map(lambda l: warc.WARCFile(fileobj=gzip.open(l))) # custom read your file
     .map(lambda f: read_file(f)) # custom parsing of your file
     .concat().take(100) # chunkify every 100 items
)

# Do something with each partition
b.map(...).concat().compute()
```

To get the partitions, you could pass the count of items and the chunksize.

Probably @eriknw can express this idea better than me.

**Comment by mrocklin** [comment]
This would let us maintain the streaming nature of things.  We would have to know how many partitions to expect ahead of time, e.g. we need to know `count(read_file(...))` ahead of time.  We would still suffer from serialization/deserialization though if you're doing time-consuming computations then maybe this isn't an issue.

**Comment by mrocklin** [comment]
In general infinite streams don't fit the dask computation model well.  We could always branch out and handle these problems, it would probably be a separate project though, not dask.

**Comment by eriknw** [comment]
I'd like to take a stab at this today or tomorrow, and I plan to support any iterable of unknown length for `dask.bag`.  Perhaps this can be `dask.bag.from_iterable`, or, as you said, broken out into a separate project or dask submodule.

I understand this doesn't fit nicely in the current dask computational model, and that reading from the iterable and serializing the elements in the main process is more likely to be the bottleneck than with typical usage of `dask.bag`.  Determining the data partitions *a priori* should be preferred.  However, `from_iterable` is common enough that I think we should offer a reasonable solution to make `dask.bag` more user-friendly while still guiding the user to use partitionable data.

This will be a good chance for me to develop some of the ideas I've been playing around with.

In the mean time, one way to use `dask.bag` with a generic stream is to do the chunking and combining yourself:
```python
it = iter(data)
for chunk in partition_all(1000, it):
    res = db.from_sequence(chunk).map(...).etc(...).compute()
    # fold results together as appropriate
```

**Comment by quasiben** [comment]
[Disco](https://github.com/discoproject/disco) implemented a `map_reader` -- essentially, this acts as function which delivers inputs to the `map` step.  This becomes especially useful when you need custom chunking.  I've used this when dealing with [BAM/SAM files](https://github.com/ContinuumIO/Examples/blob/master/1000Genomes/MapReduce_1000Genomes.py#L62-L79) -- perhaps something similar could be implemented in dask as a pre-input step.  Notice also that the [number of partitions](https://github.com/ContinuumIO/Examples/blob/master/1000Genomes/MapReduce_1000Genomes.py#L87) is explicitly defined.

I would imagine that the big single file and/or custom chunking is something that people would want.

**Comment by mrocklin** [comment]
> In the mean time, one way to use dask.bag with a generic stream is to do the chunking and combining yourself.

I pose that `cytoolz` would be a better fit than `dask.bag` in this case.

**Comment by eriknw** [comment]
> I pose that cytoolz would be a better fit than dask.bag in this case.

Agreed, unless the computation is very expensive.  Even then, using `dask.bag` as such is very awkward and probably not a very good fit, which is why I think it'd be great to have something like `dask.bag.from_iterable`.

In terms of MB/s or percentage, roughly how much of a difference is there between reading data from a file, and reading data *and* serializing it?  It'll depend on format and content, of course, but you've been spending time in this space, so what is your experience or intuition?  The advantage of serializing data is that you can send it to multiple workers.  The disadvantage is that it slows down data ingress.  I'm curious about the latter.

**Comment by eriknw** [comment]
The primary metric for determining whether `dask.bag.from_iterable` should use multiprocessing or not is whether the *computation* takes longer than *serialization* for each chunk.  I don't think this will be a difficult threshold to reach.

**Comment by mrocklin** [comment]
When dealing with typical json data I expect decompression/serialization/deserialization/serialization to dominate.  We can parallelize this though, line-delimited JSON can be partitioned among many workers.

For other formats I have no idea.  I suggest that we wait for a couple of these to come in before we invest much time here.  I would like to see a few real examples of where this is important before deciding on API and behavior.

You'll have no argument from me that `from_iterable` would be great.  I just don't currently see a clear path forward; if you do and if you are confident that you're hitting the use cases that are likely to come in then please charge ahead.

**Comment by eriknw** [comment]
I spent some time this weekend digging into this.

The approach I have in mind for dask.bag relies heavily on a couple of components I envision for an alternative implementation of dask.distributed: notably, a hub on a distributed node that receives tasks "on-the-fly" from a master coordinator such that the entire DAG need not be known by the hub (and, in this case, the coordinator) in advanced.  However, developing a second version of dask.distributed is probably not the best use of my time right now, and the patterns needed for `dask.bag.from_iterable` are more general than what easily fits into dask.bag.

So, I decided to direct my effort into a parallel pipe in `toolz.sandbox.ppipe` (still being developed off-line for a little bit longer--I dove a little too deeply into the rabbit hole).  Splitters and reducers/combiners are callables passed to `ppipe` that have `_toolz_split_` and `_toolz_combine_` attributes defined respectively (and decorators can make defining these easier).  I hope to allow for additional functionality found in other stream processing and Map-Reduce-*esque* workflows, but the proof-of-concept will try to keep the style of `toolz`, and, hence, will look a lot like `dask.bag`.

It's been fun digging into `dask.bag`.  Feel free to send tasks for it my way.

Anyway, we can probably close the original issue unless you want to keep this open to continue the discussion.  Supporting multiple partitions from a file is possible *if* the file is indexable/partitionable.  Line-delimited text is a good example of this.  Right now, an end-user probably isn't expected to make a custom partitioner.

**Comment by mrocklin** [comment]
I like the idea of a scheduler that maintains a growing DAG and caches some computed values.  I think that this would be a novel and potentially powerful contribution.  @cpcloud has talked about needing something like this as well.

**Comment by mrocklin** [comment]
As of #1116 we can partition single files when there is a clear delimiter.  I don't see a nice way to do this otherwise.  Closing for now.  Please reopen if there is a live use case that is not handled.

**Comment by timClicks** [comment]
@chdoig did you ever manage to find a dask-friendly method of parsing large WARC files? I am looking into this to process Commoncrawl data
